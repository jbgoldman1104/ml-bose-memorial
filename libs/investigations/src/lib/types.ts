import { InvestigationMetadata, Widget } from '@kleeen/types';

export interface ResolvedInvestigationMetadata extends InvestigationMetadata {
  cardLevel: number;
  otherSolutions?: string[]; // TODO: Improve this naming
}

export interface WidgetWithMetadata extends Widget {
  metadata: ResolvedInvestigationMetadata;
}
