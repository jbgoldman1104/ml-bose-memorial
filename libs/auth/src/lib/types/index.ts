export * from './auth';
export * from './auth-abstract';
export * from './auth-enums';
