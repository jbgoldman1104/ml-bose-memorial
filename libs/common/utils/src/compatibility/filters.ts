/* eslint no-console: ["error", { allow: ["warn", "error"] }] */

import {
  FilterOperation,
  FilterOperators,
  FilterQuery,
  FilterRule,
  Maybe,
  PrimitiveType,
  SingleFilterRuleValue,
} from '@kleeen/types';

import { upperCamelCase } from '../helpers';

interface Attribute {
  dataType?: string;
  name: string;
}

/** @deprecated Use this method to keep backward compatibility with filters generated before 10-2021, migrate to new filters version
 *
 * This method transform an object with filters generated after 10-2021 to filters with the shape of the previous version.
 *
 * @func transformFiltersToDeprecatedVersion
 * @param {DataListingArgs} input An object with a filters attribute to be transformed
 * @return {Filters} An object with filters using the previous shape
 *
 * * @see {@link http://LINK_TO_KLEEEN_DOCUMENTATION}
 */
export function transformFiltersToDeprecatedVersion(input: {
  attributes?: Attribute[];
  filters?: FilterQuery;
}): Maybe<Record<string, unknown>> {
  if (!input?.filters?.rules) return;

  const { attributes, filters } = input;
  const filtersInDeprecatedShape = {};

  const groupedRules = groupRulesByField(filters.rules);

  groupedRules.forEach((rule) => {
    const key = getDataType(rule, attributes);
    filtersInDeprecatedShape[key] = getFieldEntry(rule, filtersInDeprecatedShape[key]);
  });

  return filtersInDeprecatedShape;
}

//#region private methods

/**
 * Transform a rule to the previous version shape,
 * If the filter exists, tries to merge it with the corresponding transformation; for example,
 * now a timestamp rule date <= X and date >= Y has to become to:
 * date: { from: X, to: Y }
 *
 *
 * @param {FilterRule} rule to be transformed
 * @param {Record<string, unknown>} filter current filter if was previously transformed to compose accumulative changes
 * @returns {Record<string, unknown>} new filter transformed or composed from previous state
 */
function getFieldEntry(
  rule: FilterRule,
  filter: Record<string, unknown>,
): Record<string, unknown> | PrimitiveType | undefined {
  const values = Array.isArray(rule.value) ? rule.value : [rule.value];
  const [singleValue] = values;

  const getValue = (entity?: SingleFilterRuleValue) => {
    if (!entity) {
      return;
    }
    if (Object.prototype.hasOwnProperty.call(entity, 'displayValue')) {
      return entity.displayValue;
    }
    if (Object.prototype.hasOwnProperty.call(entity, 'id')) {
      return entity.id;
    }
  };

  switch (rule.operation) {
    case FilterOperation.GreaterOrEqualThan:
      if (filter && Object.prototype.hasOwnProperty.call(filter, FilterOperators.min)) {
        return {
          [FilterOperators.from]: getValue(singleValue),
          [FilterOperators.to]: filter.min,
        };
      }

      return {
        [FilterOperators.min]: getValue(singleValue),
      };

    case FilterOperation.Is:
      if (manageAsSingleAttribute(rule)) {
        return singleValue.id;
      }

      return {
        [FilterOperators.in]: values.map(getValue),
      };

    case FilterOperation.LessOrEqualThan:
      if (filter && Object.prototype.hasOwnProperty.call(filter, FilterOperators.max)) {
        return {
          [FilterOperators.from]: filter.max,
          [FilterOperators.to]: getValue(singleValue),
        };
      }

      return {
        [FilterOperators.max]: getValue(singleValue),
      };

    case FilterOperation.RelativeDate:
      return {
        [FilterOperators.relativeDate]: getValue(singleValue),
      };

    default:
      throw new Error(`the rule operation ${rule.operation} is not supported.`);
  }
}

/**
 * Tries to get the data type from the array of attributes,
 * if not found returns the field name capitalized as a fallback
 *
 * @param {string} fieldName to be search in the attributes list
 * @param attributes possible list of attributes to find the data type
 * @returns {string} data type id used as a filter key
 */
function getDataType(rule: FilterRule, attributes?: Attribute[]): string {
  const fieldName = rule.field;
  const attribute = attributes?.find((a) => a.name === fieldName);

  if (!attribute?.dataType) {
    console.warn(`the field ${fieldName} was not found.`);

    if (manageAsSingleAttribute(rule)) {
      return fieldName;
    }

    return upperCamelCase(fieldName);
  }

  return attribute.dataType;
}

/**
 * Verifies how to transform a rule,
 * one option is as a single value or as a complex object
 *
 * Example for a single value
 * Input :
 * {
 *   "field": "door",
 *   "operation": "=",
 *   "value": {
 *      "id": "25002565-c51b-4c90-b5e5-77d0941f483a"
 *   }
 *  }
 *
 * Output:
 * {
 *  "door": "4abd7d1b-5187-4a1d-bf63-e7a5daaaf226"
 * }
 *
 *
 * Example for a complex value
 * Input :
 * {
 *     "field": "alarmType",
 *     "operation": "=",
 *     "value": [
 *       {
 *         "displayValue": "DFO"
 *       },
 *       {
 *         "displayValue": "DHO"
 *       }
 *     ]
 *   }
 *
 *
 * "AlarmType": {
 *   "_in": [
 *     "DFO",
 *     "DHO"
 *   ]
 * }
 *
 * @param {FilterRule} rule to be transformed
 * @returns true if the transformation has to be as a single value, otherwise as a complex object
 */
function manageAsSingleAttribute(rule: FilterRule): boolean {
  const isSingleValue = !Array.isArray(rule.value);

  if (isSingleValue && Object.prototype.hasOwnProperty.call(rule.value, 'id')) {
    return true;
  }

  return false;
}

/**
 * Simplifies a set of rules by grouping them by field and making the value an array
 * when there are multiple rules of the same field
 *
 * Example
 *
 * [
  {
    field: 'alarmType',
    value: {
      displayValue: 'DFO',
    },
  }
  {
    field: 'alarmType',
    value: {
      displayValue: 'DHO',
    },
  }
  {
    field: 'door',
    value: {
      id: '4abd7d1b-5187-4a1d-bf63-e7a5daaaf226',
    },
  }
]

=>

[
  {
    field: 'alarmType',
    value: [
      {
        displayValue: 'DFO',
      },
      {
        displayValue: 'DHO',
      },
    ]
  },
  {
    field: 'player',
    value: {
      id: '4abd7d1b-5187-4a1d-bf63-e7a5daaaf226',
    },
  }
]
 * @param rules Source rules that can contain repeated fields
 * @returns The same rules but grouped by field
 */
function groupRulesByField(rules: FilterRule[]): FilterRule[] {
  const groupedRulesMap = rules.reduce((acc, rule) => {
    const currentRule = acc[rule.field];
    if (currentRule) {
      if (!Array.isArray(currentRule.value)) {
        const value = [
          {
            ...currentRule.value,
          },
          {
            ...rule.value,
          },
        ];

        acc[rule.field] = {
          ...rule,
          value,
        };

        return acc;
      } else {
        acc[rule.field] = {
          ...rule,
          value: [...currentRule, rule.value],
        };

        return acc;
      }
    }

    acc[rule.field] = rule;
    return acc;
  }, {});

  return Object.values(groupedRulesMap);
}

//#endregion private methods
