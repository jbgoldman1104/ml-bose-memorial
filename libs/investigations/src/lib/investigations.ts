import {
  Filters,
  InvestigationCard,
  InvestigationDataPoint,
  InvestigationMetadata,
  WidgetTypes,
} from '@kleeen/types';

export interface GetInitialInvestigation {
  investigationDataPoint: InvestigationDataPoint;
  investigationFilters: Filters;
  originWidgetChartType?: WidgetTypes;
  originWidgetId?: string;
  pageFilters: Filters;
  user?: string;
}

export function getInitialInvestigationCards({
  investigationDataPoint,
  investigationFilters,
  originWidgetChartType,
  originWidgetId,
  pageFilters,
  user,
}: GetInitialInvestigation): InvestigationCard {
  const metadata: InvestigationMetadata = {
    createdAt: new Date().toISOString(),
    user,
  };

  const firstInvestigationCard: InvestigationCard = {
    dataPoint: investigationDataPoint,
    filters: { ...pageFilters, ...investigationFilters },
    metadata,
  };
  const result: InvestigationCard = {
    filters: pageFilters,
    followUpCards: [firstInvestigationCard],
    metadata,
    widgetId: originWidgetId,
    widgetChartType: originWidgetChartType,
  };

  return result;
}

export interface GetInvestigationCardByDataPoint {
  inheritedFilters: Filters;
  dataPoint: InvestigationDataPoint;
  user?: string;
}

export function getInvestigationCardByDataPoint({
  dataPoint,
  inheritedFilters,
  user,
}: GetInvestigationCardByDataPoint): InvestigationCard {
  return {
    dataPoint,
    filters: inheritedFilters,
    followUpCards: [],
    metadata: {
      createdAt: new Date().toISOString(),
      user,
    },
  };
}
