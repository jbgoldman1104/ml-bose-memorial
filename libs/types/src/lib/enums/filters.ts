export enum FilterCombinator {
  AND = 'AND',
  OR = 'OR',
}

export enum FilterOperation {
  GreaterOrEqualThan = '>=',
  Is = '=',
  LessOrEqualThan = '<=',
  RelativeDate = 'relativeDate',
}

/**
 * @deprecated This will be deprecated in coming Releases, please only use it for testing purposes, use FilterOperation instead.
 */
export enum FilterOperators {
  from = 'from',
  in = '_in',
  max = 'max',
  min = 'min',
  relativeDate = 'relativeDate',
  to = 'to',
}
