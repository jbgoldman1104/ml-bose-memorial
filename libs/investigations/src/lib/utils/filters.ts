import { Filters, Maybe, Transformation, Widget, WidgetScope } from '@kleeen/types';
import {
  GetContextDataPointFiltersArgs,
  GetContextFilters,
  GetDataPointFilters,
  GetFiltersForDataPointsArgs,
  GetWidgetWithFilters,
  GetWidgetsWithFiltersProps,
  MapFilterToWidgetProps,
  WidgetsWithFilters,
} from './filters.types';

import { isSingleCardinalityTransformation } from '@kleeen/frontend/utils';

export interface FilterVariants {
  contextDataPointFilters: Maybe<Filters>;
  contextFilters: Maybe<Filters>;
  dataPointFilters: Maybe<Filters>;
}

export function getFiltersForDataPoints({
  contextDataPoints,
  dataPoint,
  scope,
  widgetContextParams,
}: GetFiltersForDataPointsArgs): FilterVariants {
  return {
    contextDataPointFilters: getContextDataPointFilters({ contextDataPoints }),
    contextFilters: getContextFilters({ widgetContextParams }),
    dataPointFilters: getDataPointFilters({ dataPoint, scope }),
  };
}

export function getWidgetsDecoratedWithFilters({
  widgets,
  ...rest
}: GetWidgetsWithFiltersProps): WidgetsWithFilters {
  const { contextDataPointFilters, contextFilters, dataPointFilters } = getFiltersForDataPoints(rest);
  return {
    widgetsWithContextDataPointFilters: widgets.map(
      mapFilterToWidget({
        filtersToApply: {
          ...contextDataPointFilters,
          ...contextFilters,
          ...dataPointFilters,
        },
      }),
    ),
    widgetsWithDefaultFilters: widgets.map(
      mapFilterToWidget({
        filtersToApply: {
          ...contextFilters,
          ...dataPointFilters,
        },
      }),
    ),
  };
}

export function getWidgetWithFilters({
  filters,
  overridePreviousFilters = false,
  widget,
}: GetWidgetWithFilters): Widget {
  return {
    ...widget,
    params: {
      ...widget?.params,
      filters: {
        ...(overridePreviousFilters ? {} : widget?.params?.filters),
        ...filters,
      },
    },
  };
}

//#region Private Members

function getContextDataPointFilters({ contextDataPoints }: GetContextDataPointFiltersArgs): Maybe<Filters> {
  const filters = contextDataPoints
    .filter((dataPoint) => {
      return isSingleCardinalityTransformation(dataPoint.attribute.aggregation as Transformation);
    })
    .reduce((acc: Filters, dataPoint) => {
      // TODO: @cafe This filter composition may change after the filters refactor
      acc[dataPoint.attribute.name] = dataPoint.value?.id || dataPoint.value?.displayValue;

      return acc;
    }, {});

  return filters;
}

function getContextFilters({ widgetContextParams }: GetContextFilters): Maybe<Filters> {
  return widgetContextParams?.filters;
}

function getDataPointFilters({ dataPoint, scope }: GetDataPointFilters): Maybe<Filters> {
  if (scope !== WidgetScope.Single) {
    return;
  }
  return {
    [dataPoint.attribute.name]: dataPoint.value.id,
  };
}

function mapFilterToWidget({ filtersToApply }: MapFilterToWidgetProps): (widget: Widget) => Widget {
  return (widget: Widget) => {
    return getWidgetWithFilters({
      widget,
      filters: filtersToApply,
    });
  };
}
//#endregion
