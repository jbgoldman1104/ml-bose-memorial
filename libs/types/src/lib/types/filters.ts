import { AccessControlPermission, FilterCombinator, FilterOperation, Transformation } from '../enums';

import { PrimitiveType } from './base';

export type Filters = Record<string, any>;

export interface SingleFilterRuleValue {
  displayValue?: PrimitiveType;
  id?: PrimitiveType;
}

export interface FilterRule {
  combinator?: FilterCombinator;
  field: string;
  operation: FilterOperation;
  transformation?: Transformation;
  value: FilterRuleValue;
}

export type FilterRuleValue = SingleFilterRuleValue | SingleFilterRuleValue[];

export interface FilterQuery {
  combinator: FilterCombinator;
  rules: FilterRule[];
}

export interface WorkflowFilter {
  accessLevel: AccessControlPermission;
  name: string;
  statisticalType: string;
}
