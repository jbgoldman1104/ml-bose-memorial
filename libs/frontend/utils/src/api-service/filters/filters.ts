import {
  FilterCombinator,
  FilterOperation,
  FilterOperators,
  FilterQuery,
  FilterRule,
  Filters,
  Maybe,
  ParamsBasedOnRoute,
  PrimitiveType,
} from '@kleeen/types';

import { INVESTIGATION_URL_PARAM } from '@kleeen/investigations';
import camelCase from 'lodash.camelcase';
import { isNilOrEmpty } from '@kleeen/common/utils';

const operationMap: { [key in FilterOperators]: FilterOperation } = {
  [FilterOperators.from]: FilterOperation.GreaterOrEqualThan,
  [FilterOperators.in]: FilterOperation.Is,
  [FilterOperators.max]: FilterOperation.LessOrEqualThan,
  [FilterOperators.min]: FilterOperation.GreaterOrEqualThan,
  [FilterOperators.relativeDate]: FilterOperation.RelativeDate,
  [FilterOperators.to]: FilterOperation.LessOrEqualThan,
};

export function getFiltersInput(filters?: Filters): FilterQuery | Record<string, never> {
  if (isNilOrEmpty(filters)) return {};

  const rules = Object.entries(filters).reduce((acc: FilterRule[], [thingName, operation]) => {
    if (thingName === INVESTIGATION_URL_PARAM) {
      // Ignore the special Investigation param
      return acc;
    }

    const field = getFilterFieldName(thingName);
    const isPrimitive =
      typeof operation == 'boolean' || typeof operation == 'number' || typeof operation == 'string';

    if (isPrimitive) {
      const ruleFromPrimitive = {
        field,
        operation: FilterOperation.Is,
        value: { id: operation },
      };

      acc.push(ruleFromPrimitive);

      return acc;
    }

    const filterRulesFromObject = getFilterOperationsAndValues({
      field,
      operation,
    });

    if (isNilOrEmpty(filterRulesFromObject)) return acc;

    acc.push(...filterRulesFromObject);

    return acc;
  }, []);
  const filtersInput = {
    combinator: FilterCombinator.AND,
    rules,
  };

  return filtersInput;
}

//#region Private members
interface GetFilterOperationsAndValues {
  field: string;
  operation: ParamsBasedOnRoute;
}

function getFilterOperationsAndValues({
  field,
  operation,
}: GetFilterOperationsAndValues): Maybe<FilterRule[]> {
  const operationObject = Object.entries(operation);

  if (isNilOrEmpty(operationObject)) return;

  const filterOperationsAndValues = operationObject.reduce((acc: FilterRule[], [oldOperator, value]) => {
    if (isNilOrEmpty(value)) return acc;

    const baseFilterOperationAndValue = {
      field,
      operation: operationMap[oldOperator as FilterOperators],
    };

    if (Array.isArray(value)) {
      const displayValues = value as PrimitiveType[];
      const rules = displayValues.map((displayValue): FilterRule => {
        return {
          ...baseFilterOperationAndValue,
          value: {
            displayValue,
          },
        };
      });

      acc.push(...rules);

      return acc;
    }

    const displayValue = value as PrimitiveType;
    const filterOperationAndValue = {
      ...baseFilterOperationAndValue,
      value: { displayValue },
    };

    acc.push(filterOperationAndValue);

    return acc;
  }, []);

  return filterOperationsAndValues;
}

function getFilterFieldName(originalName: string): string {
  return camelCase(originalName);
}
//#endregion
