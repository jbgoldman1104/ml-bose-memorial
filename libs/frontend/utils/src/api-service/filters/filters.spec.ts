/* eslint-disable max-lines */
import { FilterCombinator, FilterOperation, FilterOperators, IntervalDate } from '@kleeen/types';

import { INVESTIGATION_URL_PARAM } from '@kleeen/investigations';
import { getFiltersInput } from './filters';

describe('api-service', () => {
  describe('filters', () => {
    it('should skip adding rules objects if "paramsBasedOnRoute" has incomplete operations', () => {
      const filterMap = {
        ComputerChild: {},
        ComputerParent: {
          [FilterOperators.in]: [],
        },
      };
      const filterQuery = {
        combinator: FilterCombinator.AND,
        rules: [],
      };

      expect(getFiltersInput(filterMap)).toEqual(filterQuery);
    });

    it('should ignore the investigation special param "i"', () => {
      const filterMap = {
        computerChild: '39a91f5e-da34-4e64-bd0f-730a7c9d6a99',
        computerParent: 'a7e2e0aa-7bf5-4113-8139-7167fe1b8770',
        [INVESTIGATION_URL_PARAM]:
          'eyJmb2xsb3dVcENhcmRzIjpbeyJkYXRhUG9pbnQiOnsiZW50aXR5SWQiOjE3NTM0Nywic2NvcGUiOiJzaW5nbGUifSwibWV0YWRhdGEiOnsiY3JlYXRlZEF0IjoiMjAyMS0xMC0xMlQyMDo1OTozNC41NThaIn0sImZpbHRlcnMiOnsicGxheWVyIjoiZjljODhiYzItNGU1MC00MzIzLTkwNDUtNzE1NzRiZjI2ZmIzIn19XSwibWV0YWRhdGEiOnsiY3JlYXRlZEF0IjoiMjAyMS0xMC0xMlQyMDo1OTozNC41NThaIn0sImZpbHRlcnMiOnt9fQ',
      };
      const filterQuery = {
        combinator: FilterCombinator.AND,
        rules: [
          {
            field: 'computerChild',
            operation: FilterOperation.Is,
            value: { id: '39a91f5e-da34-4e64-bd0f-730a7c9d6a99' },
          },
          {
            field: 'computerParent',
            operation: FilterOperation.Is,
            value: { id: 'a7e2e0aa-7bf5-4113-8139-7167fe1b8770' },
          },
        ],
      };

      expect(getFiltersInput(filterMap)).toEqual(filterQuery);
    });

    it('should return an empty filters object input if "paramsBasedOnRoute" is nilOrEmpty', () => {
      const filterQuery = {};

      expect(getFiltersInput({})).toEqual(filterQuery);
      expect(getFiltersInput()).toEqual(filterQuery);
      expect(getFiltersInput(undefined)).toEqual(filterQuery);
    });

    it(`should return a "${FilterOperation.Is}" operation if "paramsBasedOnRoute" has a plain parameter operation`, () => {
      const filterMap = {
        computerChild: '39a91f5e-da34-4e64-bd0f-730a7c9d6a99',
        computerParent: 12345,
        isParent: false,
      };
      const filterQuery = {
        combinator: FilterCombinator.AND,
        rules: [
          {
            field: 'computerChild',
            operation: FilterOperation.Is,
            value: { id: '39a91f5e-da34-4e64-bd0f-730a7c9d6a99' },
          },
          {
            field: 'computerParent',
            operation: FilterOperation.Is,
            value: { id: 12345 },
          },
          {
            field: 'isParent',
            operation: FilterOperation.Is,
            value: { id: false },
          },
        ],
      };

      expect(getFiltersInput(filterMap)).toEqual(filterQuery);
    });

    it(`should return a "${FilterOperation.Is}" operation if "paramsBasedOnRoute" has a "${FilterOperators.in}" operation`, () => {
      const filterMap = {
        Computer: {
          [FilterOperators.in]: [12345],
        },
        ComputerChild: {
          [FilterOperators.in]: ['quia aut dolorum at porro deserunt accusamus'],
        },
        ComputerParent: {
          [FilterOperators.in]: ['optio nobis maiores', 'quisquam voluptate molestias eius'],
        },
      };
      const filterQuery = {
        combinator: FilterCombinator.AND,
        rules: [
          {
            field: 'computer',
            value: { displayValue: 12345 },
            operation: FilterOperation.Is,
          },
          {
            field: 'computerChild',
            value: { displayValue: 'quia aut dolorum at porro deserunt accusamus' },
            operation: FilterOperation.Is,
          },
          {
            field: 'computerParent',
            value: { displayValue: 'optio nobis maiores' },
            operation: FilterOperation.Is,
          },
          {
            field: 'computerParent',
            value: { displayValue: 'quisquam voluptate molestias eius' },
            operation: FilterOperation.Is,
          },
        ],
      };

      expect(getFiltersInput(filterMap)).toEqual(filterQuery);
    });

    it(`should return a "${FilterOperation.LessOrEqualThan}" operation if "paramsBasedOnRoute" has a "${FilterOperators.max}" operation`, () => {
      const filterMap = {
        Salary: {
          [FilterOperators.max]: 80,
        },
      };
      const filterQuery = {
        combinator: FilterCombinator.AND,
        rules: [
          {
            field: 'salary',
            value: { displayValue: 80 },
            operation: FilterOperation.LessOrEqualThan,
          },
        ],
      };

      expect(getFiltersInput(filterMap)).toEqual(filterQuery);
    });

    it(`should return a "${FilterOperation.GreaterOrEqualThan}" operation if "paramsBasedOnRoute" has a "${FilterOperators.min}" operation`, () => {
      const filterMap = {
        Salary: {
          [FilterOperators.min]: -45,
        },
      };
      const filterQuery = {
        combinator: FilterCombinator.AND,
        rules: [
          {
            field: 'salary',
            value: { displayValue: -45 },
            operation: FilterOperation.GreaterOrEqualThan,
          },
        ],
      };

      expect(getFiltersInput(filterMap)).toEqual(filterQuery);
    });

    it(`should return a range operation if "paramsBasedOnRoute" has both "${FilterOperators.min}" and "${FilterOperators.max}" operations`, () => {
      const filterMap = {
        Salary: {
          [FilterOperators.max]: 80,
          [FilterOperators.min]: -45,
        },
      };
      const filterQuery = {
        combinator: FilterCombinator.AND,
        rules: [
          {
            field: 'salary',
            value: { displayValue: 80 },
            operation: FilterOperation.LessOrEqualThan,
          },
          {
            field: 'salary',
            value: { displayValue: -45 },
            operation: FilterOperation.GreaterOrEqualThan,
          },
        ],
      };

      expect(getFiltersInput(filterMap)).toEqual(filterQuery);
    });

    it(`should return a range operation if "paramsBasedOnRoute" has both "${FilterOperators.from}" and "${FilterOperators.to}" operations`, () => {
      const filterMap = {
        Timestamp: {
          [FilterOperators.from]: 1630497600000,
          [FilterOperators.to]: 1632794400000,
        },
      };
      const filterQuery = {
        combinator: FilterCombinator.AND,
        rules: [
          {
            field: 'timestamp',
            value: { displayValue: 1630497600000 },
            operation: FilterOperation.GreaterOrEqualThan,
          },
          {
            field: 'timestamp',
            value: { displayValue: 1632794400000 },
            operation: FilterOperation.LessOrEqualThan,
          },
        ],
      };

      expect(getFiltersInput(filterMap)).toEqual(filterQuery);
    });

    it.each([
      [IntervalDate.allTime, IntervalDate.allTime],
      [IntervalDate.minute, IntervalDate.minute],
      [IntervalDate.oneHours, IntervalDate.oneHours],
      [IntervalDate.oneMonth, IntervalDate.oneMonth],
      [IntervalDate.oneWeek, IntervalDate.oneWeek],
      [IntervalDate.sixHours, IntervalDate.sixHours],
      [IntervalDate.threeMonth, IntervalDate.threeMonth],
      [IntervalDate.twentyFourHours, IntervalDate.twentyFourHours],
    ])(
      `should return a "${FilterOperation.RelativeDate}" of %p if "paramsBasedOnRoute" has a "${FilterOperators.relativeDate}" of %p operation`,
      (interval, expectedInterval) => {
        const filterMap = {
          Timestamp: {
            [FilterOperators.relativeDate]: interval,
          },
        };
        const filterQuery = {
          combinator: FilterCombinator.AND,
          rules: [
            {
              field: 'timestamp',
              value: { displayValue: expectedInterval },
              operation: FilterOperation.RelativeDate,
            },
          ],
        };

        expect(getFiltersInput(filterMap)).toEqual(filterQuery);
      },
    );

    it(`should return a combination of all operations if "paramsBasedOnRoute" has "${FilterOperators.in}", "${FilterOperators.max}", "${FilterOperators.min}", "${FilterOperators.from}" and "${FilterOperators.to}" operations`, () => {
      const filterMap = {
        CoffeeShop: {
          [FilterOperators.in]: ['Chocolatito'],
        },
        Salary: {
          [FilterOperators.in]: 20,
          [FilterOperators.max]: 23,
          [FilterOperators.min]: 14,
        },
        Timestamp: {
          [FilterOperators.from]: 1633101300000,
          [FilterOperators.to]: 1633274100000,
        },
      };

      expect(getFiltersInput(filterMap)).toMatchSnapshot();
    });
  });
});
