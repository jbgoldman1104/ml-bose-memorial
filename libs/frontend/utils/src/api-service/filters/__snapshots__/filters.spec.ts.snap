// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api-service filters should return a combination of all operations if "paramsBasedOnRoute" has "_in", "max", "min", "from" and "to" operations 1`] = `
Object {
  "combinator": "AND",
  "rules": Array [
    Object {
      "field": "coffeeShop",
      "operation": "=",
      "value": Object {
        "displayValue": "Chocolatito",
      },
    },
    Object {
      "field": "salary",
      "operation": "=",
      "value": Object {
        "displayValue": 20,
      },
    },
    Object {
      "field": "salary",
      "operation": "<=",
      "value": Object {
        "displayValue": 23,
      },
    },
    Object {
      "field": "salary",
      "operation": ">=",
      "value": Object {
        "displayValue": 14,
      },
    },
    Object {
      "field": "timestamp",
      "operation": ">=",
      "value": Object {
        "displayValue": 1633101300000,
      },
    },
    Object {
      "field": "timestamp",
      "operation": "<=",
      "value": Object {
        "displayValue": 1633274100000,
      },
    },
  ],
}
`;
